### PR TITLE
Support .openpublishing.redirection.json

### PR DESCRIPTION
**Pull request by yufeih**
Support `.openpublishing.redirection.json` config at repo root.

For the existing config, moved `redirections` + `redirectionsWithoutId` outside `docfx.yml` to a dedicated file at docset path (`redirections.yml`). There is no such file in DHS so we can safely specialize this file name. This file shares the same schema as `.openpublishing.redirection.json`. A shorthand object form is added to make it less verbose. `redirections/renames` pair is chosen over the existing `redirectionsWithoutId/redirections` pair.

`source_path` is defined and rebased relative to the location of declaring file.

Source info can be reported correct for `.openpublishing.redirection.json` now.
 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5249)

### CODE REVIEW

**Comment by yufeih** [comment]
hmmm https://github.com/MicrosoftDocs/docs-help-pr/blob/master/help-content/onboard/admin/reference/dotnet/redirection.md

File: docs/specs/redirection.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        {
+          "redirections": [
+            {"source_path": "index.md", "redirect_url": "/"},
+            {"source_path": "a/index.md", "redirect_url": "/a"},
+          ]
+        }
+      a/docfx.yml:
+outputs:
+  a/.publish.json: |
+    { "files": [{ "source_path": "index.md", "redirect_url": "/a" }] }
```
Comment (928PJY) [inline docs/specs/redirection.yml]: add a error test?

**Comment by mingwli** [comment]
```diff
+  "file": "microsoft-edge/../.openpublishing.redirection.json",
```
seems this should be normalized in adjust path.

File: src/docfx/docset/Document.cs [hunk 2]
```diff
@@ -391,7 +395,7 @@ private static bool IsValidRelativePath(string path)
                 : mappedSourcePath;
 
             // if source is landing page, change it to *.md
```
Comment (mingwli) [inline src/docfx/docset/Document.cs]: the comment is out-of-date too

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 3]
```diff
@@ -51,21 +52,26 @@ public bool TryGetDocumentId(Document file, out (string id, string versionIndepe
...
-            AddRedirections(docset.Config.RedirectionsWithoutId);
+            AddRedirections(redirectionItems.Where(item => !item.RedirectDocumentId));
 
             var redirectionsBySourcePath = redirections.ToDictionary(file => file.FilePath.Path, file => file, PathUtility.PathComparer);
             var redirectionsByTargetSourcePath = GetRedirectionsByTargetSourcePath(errorLog, redirectionsWithDocumentId, buildFiles.Concat(redirections).ToList(), monikerProvider);
 
             return new RedirectionMap(redirectionsBySourcePath, redirectionsByTargetSourcePath);
 
-            void AddRedirections(Dictionary<string, SourceInfo<string>> items, bool redirectDocumentId = false)
+            void AddRedirections(IEnumerable<RedirectionItem> items, bool redirectDocumentId = false)
```
Comment (mingwli) [inline src/docfx/build/redirection/RedirectionMap.cs]: >bool redirectDocumentId = false [](start = 69, length = 31)

can `redirectDocumentId` paramete be removed?

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 4]
```diff
@@ -51,21 +52,26 @@ public bool TryGetDocumentId(Document file, out (string id, string versionIndepe
             var redirections = new HashSet<Document>();
             var redirectionsWithDocumentId = new List<(SourceInfo<string> originalRedirectUrl, Document redirect)>();
 
+            var redirectionItems = LoadRedirectionModel(docset.DocsetPath);
+
             // load redirections with document id
-            AddRedirections(docset.Config.Redirections, redirectDocumentId: true);
+            AddRedirections(redirectionItems.Where(item => item.RedirectDocumentId), redirectDocumentId: true);
```
Comment (mingwli) [inline src/docfx/build/redirection/RedirectionMap.cs]:
```suggestion
            AddRedirections(redirectionItems);
```

**Comment by yufeih** [comment]
> ```diff
> +  "file": "microsoft-edge/../.openpublishing.redirection.json",
> ```
> 
> seems this should be normalized in adjust path.

@live1206 , this is due to post process in docs-build, docfx outputs normalized path

**Comment by live1206** [comment]
> > ```diff
> > +  "file": "microsoft-edge/../.openpublishing.redirection.json",
> > ```
> > 
> > 
> > seems this should be normalized in adjust path.
> 
> @live1206 , this is due to post process in docs-build, docfx outputs normalized path

https://ceapex.visualstudio.com/Engineering/_git/Docs.Build/pullrequest/17840?_a=overview

**Pull request merged**